src/hazelcore/Statics_Exp.re

```diff
@@ -121,95 +123,91 @@ and syn_skel =
 and syn_operand = (ctx: Contexts.t, operand: UHExp.operand): option(HTyp.t) =>
   switch (operand) {
   /* in hole */
-  | EmptyHole(_) => Some(Hole)
-  | InvalidText(_) => Some(Hole)
+  | EmptyHole(_) => Some(Unknown(Internal))
+  | InvalidText(_) => Some(Unknown(Internal))
   | Var(InHole(TypeInconsistent, _), _, _)
   | IntLit(InHole(TypeInconsistent, _), _)
   | FloatLit(InHole(TypeInconsistent, _), _)
   | BoolLit(InHole(TypeInconsistent, _), _)
   | ListNil(InHole(TypeInconsistent, _))
   | Fun(InHole(TypeInconsistent, _), _, _)
-  | Inj(InHole(TypeInconsistent, _), _, _)
-  | Case(StandardErrStatus(InHole(TypeInconsistent, _)), _, _) =>
+  | Inj(InHole(TypeInconsistent, _), _, _) =>
     let operand' = UHExp.set_err_status_operand(NotInHole, operand);
     let+ _ = syn_operand(ctx, operand');
-    HTyp.Hole;
+    HTyp.Unknown(Internal);
   | Var(InHole(WrongLength, _), _, _)
   | IntLit(InHole(WrongLength, _), _)
   | FloatLit(InHole(WrongLength, _), _)
   | BoolLit(InHole(WrongLength, _), _)
   | ListNil(InHole(WrongLength, _))
   | Fun(InHole(WrongLength, _), _, _)
-  | Inj(InHole(WrongLength, _), _, _)
-  | Case(StandardErrStatus(InHole(WrongLength, _)), _, _) => None
-  | Case(InconsistentBranches(rule_types, _), scrut, rules) =>
-    let* pat_ty = syn(ctx, scrut);
-    /* Make sure the rule synthesizes the type the rule_types says it does */
-    let correct_rule_types =
-      List.for_all2(
-        (rule_ty, rule) => {
-          switch (syn_rule(ctx, rule, pat_ty)) {
-          | None => false
-          | Some(syn_ty) => HTyp.eq(rule_ty, syn_ty)
-          }
-        },
-        rule_types,
-        rules,
-      );
-    correct_rule_types ? Some(HTyp.Hole) : None;
+  | Inj(InHole(WrongLength, _), _, _) => None
+  | Case(InconsistentBranches(_, Syn), scrut, rules) =>
+    let* ty_scrut = syn(ctx, scrut);
+    let* clause_types = syn_rule_types(ctx, ty_scrut, rules);
+    switch (HTyp.join_all(GLB, clause_types)) {
+    | None => Some(HTyp.Unknown(Internal))
+    | Some(_) => None
+    };
+  | Case(InconsistentBranches(_, Ana), scrut, _) =>
+    let* _ty_scrut = syn(ctx, scrut);
```

Comment:
I think we do need to check that all of the rules synthesize a type, just don't need to do the join here.